sw/device/tests/crypto/aes_gcm_functest.c

```diff
@@ -0,0 +1,184 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/crypto/aes_gcm/aes_gcm.h"
+#include "sw/device/lib/crypto/drivers/aes.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+typedef struct aes_gcm_test {
+  // Key length.
+  aes_key_len_t key_len;
+  // IV and length (in bytes). If IV length is < 16 then the last bytes are
+  // ignored.
+  size_t iv_len;
+  uint8_t iv[16];
+  // Plaintext and length (in bytes). If the length is not a multiple of 4,
+  // then the most significant bytes of the last word are ignored.
+  size_t plaintext_len;
+  uint8_t *plaintext;
+  // Authenticated data and length (in bytes). If the length is not a multiple
+  // of 4, then the most significant bytes of the last word are ignored.
+  size_t aad_len;
+  uint8_t *aad;
+  // Expected authentication tag.
+  uint8_t expected_tag[16];
+  // Expected ciphertext (same length as plaintext).
+  uint8_t *expected_ciphertext;
+} aes_gcm_test_t;
+
+/**
+ * Randomly-generated 128-bit key for testing.
+ */
+static uint32_t kKey128[4] = {
+    // Key = f80a6e67211c873793a99d899c31c2e7
+    0x676e0af8, 0x37871c21, 0x899da993, 0xe7c2319c};
+
+/**
+ * Randomly-generated 256-bit key for testing.
+ */
+static uint32_t kKey256[8] = {
+    // Key = 76592790eaf6630e670ce5784ff23a1806a1ea76b0977b1542374769247cc4ce
+    0x90275976, 0x0e63f6ea, 0x78e50c67, 0x183af24f,
+    0x76eaa106, 0x157b97b0, 0x69473742, 0xcec47c24};
+
+/**
+ * Authenticated data for testing.
+ */
+static const uint32_t kAadLen = 18;
+static uint8_t kAad[20] = {
+    // aad = 'authenticated data'
+    //     = 61757468656e746963617465642064617461
+    0x61, 0x75, 0x74, 0x68, 0x65, 0x6e, 0x74, 0x69, 0x63,
+    0x61, 0x74, 0x65, 0x64, 0x20, 0x64, 0x61, 0x74, 0x61};
+
+/**
+ * Plaintext for testing.
+ */
+static const uint32_t kPlaintextLen = 32;
+static uint8_t kPlaintext[32] = {
+    // plaintext = 'authenticated and encrypted data'
+    //           =
+    //           61757468656e7469636174656420616e6420656e637279707465642064617461
+    0x61, 0x75, 0x74, 0x68, 0x65, 0x6e, 0x74, 0x69, 0x63, 0x61, 0x74,
+    0x65, 0x64, 0x20, 0x61, 0x6e, 0x64, 0x20, 0x65, 0x6e, 0x63, 0x72,
+    0x79, 0x70, 0x74, 0x65, 0x64, 0x20, 0x64, 0x61, 0x74, 0x61};
+
+/**
+ * Expected ciphertext for the 256-bit key.
+ */
+static uint8_t kCiphertext256[32] = {
+    // Ciphertext =
+    // 4e6d3a963b076ba0945d29aa836f29b0fa06cdd575aab8233f1df93e80163371
+    0x4e, 0x6d, 0x3a, 0x96, 0x3b, 0x07, 0x6b, 0xa0, 0x94, 0x5d, 0x29,
+    0xaa, 0x83, 0x6f, 0x29, 0xb0, 0xfa, 0x06, 0xcd, 0xd5, 0x75, 0xaa,
+    0xb8, 0x23, 0x3f, 0x1d, 0xf9, 0x3e, 0x80, 0x16, 0x33, 0x71};
+
+static const aes_gcm_test_t kAesGcmTestvectors[3] = {
+    // Empty input, empty aad, 12-bit IV, 128-bit key
```

Comment:
Nit: `96-bit IV`, right? I think the other IV lengths in comments are mislabeled as bits as well :)